sdks/python/apache_beam/coders/coders.py

```
@@ -1383,22 +1375,74 @@ def from_runner_api_parameter(payload, components, context):
 
-  def proto(self):
-    return self._proto
+  coder_count = 0
 
-  def to_runner_api(self, context):
-    return self._proto
+  def __init__(self, element_type_holder):
+    self.element_type_holder = element_type_holder
 
-  def to_type_hint(self):
-    return Any
+  def as_cloud_object(self, coders_context=None):
+    if not coders_context:
+      raise Exception(
+          'coders_context must be specified to correctly encode external coders'
+      )
+    coder_id = coders_context.get_by_proto(
+        self.element_type_holder.coder, deduplicate=True)
+
+    coder_proto = self.element_type_holder.coder
+
+    kind_str = 'kind:external' + str(ExternalCoder.coder_count)
+    ExternalCoder.coder_count = ExternalCoder.coder_count + 1
+    component_encodings = []
+    if coder_proto.spec.urn == 'beam:coder:kv:v1':
+      kind_str = 'kind:pair'
+      for component_coder_id in coder_proto.component_coder_ids:
+        component_encodings.append({
+            '@type': 'kind:external' + str(ExternalCoder.coder_count),
+            'pipeline_proto_coder_id': component_coder_id
+        })
+        ExternalCoder.coder_count = ExternalCoder.coder_count + 1
+
+    value = {
+        # This is a placeholder type. Dataflow will get the actual coder from
+        # pipeline proto using the pipeline_proto_coder_id property.
+        '@type': kind_str,
+        'pipeline_proto_coder_id': coder_id
+    }
+    if component_encodings:
+      value['is_pair_like'] = True
+      value['component_encodings'] = component_encodings
+
+    return value
+
+  @staticmethod
+  def from_type_hint(typehint, unused_registry):
+    if isinstance(typehint, ElementTypeHolder):
+      return ExternalCoder(typehint)
+    else:
+      raise ValueError((
+          'Expected an instance of ElementTypeHolder'
+          ', but got a %s' % typehint))
+
+  def to_runner_api_parameter(self, context):
+    if self.element_type_holder.coder.component_coder_ids:
+      raise NotImplementedError
```

Why raise NotImplementedError here instead of returning this in place of the empty tuple below?